controller/openebs/cstor.go

```diff
@@ -111,6 +150,532 @@ func (p *Planner) setCStorDefaultsIfNotSet() error {
+		}
+		if vmName == "iscsiadm-bin" {
+			err = unstructured.SetNestedField(vm.Object,
+				p.ObservedOpenEBS.Spec.CstorConfig.CStorCSI.CStorCSINode.ISCSIPath, "spec", "mountPath")
+			if err != nil {
+				return err
+			}
+		}
+		return nil
+	}
+
+	// update the containers
+	updateContainer := func(obj *unstructured.Unstructured) error {
+		containerName, _, err := unstructured.NestedString(obj.Object, "spec", "name")
+		if err != nil {
+			return err
+		}
+		envs, _, err := unstruct.GetSlice(obj, "spec", "env")
+		if err != nil {
+			return err
+		}
+		volumeMounts, _, err := unstruct.GetSlice(obj, "spec", "volumeMounts")
+		if err != nil {
+			return err
+		}
+
+		if containerName == ContainerOpenEBSCSIPluginName {
+			// Set the image of the container.
+			err = unstructured.SetNestedField(obj.Object, p.ObservedOpenEBS.Spec.CstorConfig.CStorCSI.CStorCSINode.Image,
+				"spec", "image")
+			if err != nil {
+				return err
+			}
+			// Set the environmets of the container.
+			err = unstruct.SliceIterator(envs).ForEachUpdate(updateOpenEBSCSIPluginEnv)
+			if err != nil {
+				return err
+			}
+			err = unstruct.SliceIterator(volumeMounts).ForEachUpdate(updateOpenEBSCSIPluginVolumeMount)
+			if err != nil {
+				return err
+			}
+		}
+		err = unstructured.SetNestedSlice(obj.Object, envs, "spec", "env")
+		if err != nil {
+			return err
+		}
+		err = unstructured.SetNestedSlice(obj.Object, volumeMounts, "spec", "volumeMounts")
+		if err != nil {
+			return err
+		}
+
+		return nil
+	}
+
+	// Update the containers.
+	err = unstruct.SliceIterator(containers).ForEachUpdate(updateContainer)
+	if err != nil {
+		return err
+	}
+
+	// Set back the value of the containers.
+	err = unstructured.SetNestedSlice(daemonset.Object,
+		containers, "spec", "template", "spec", "containers")
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// addExtraVolumes will add the extra volumes in openebs-csi-plugin container for the csi to work.
+// This function will fetch/detect the OS used to install CSI components and will add the volumes depending on the OS.
+// Volumes are different for different OS.
+func (p *Planner) addExtraVolumes(daemonset *unstructured.Unstructured) error {
+	osImage, err := k8s.GetOSImage()
+	if err != nil {
+		return errors.Errorf("Error getting OS Image of a Node, error: %+v", err)
+	}
+
+	ubuntuVersion, err := k8s.GetUbuntuVersion()
+	if err != nil {
+		return errors.Errorf("Error getting Ubuntu Version of a Node, error: %+v", err)
+	}
+
+	switch true {
+	case strings.Contains(strings.ToLower(osImage), strings.ToLower(types.OSImageSLES12)):
+		err = p.addExtraVolumeMountsForSUSE12(daemonset)
+	case strings.Contains(strings.ToLower(osImage), strings.ToLower(types.OSImageSLES15)):
+		err = p.addExtraVolumeMountsForSUSE15(daemonset)
+	case strings.Contains(strings.ToLower(osImage), strings.ToLower(types.OSImageUbuntu1804)):
+		err = p.addExtraVolumeMountsForUbuntu18(daemonset)
+	case (ubuntuVersion != 0) && ubuntuVersion >= 18.04:
```

Comment:
this case will take care of the above one also, and aren't we concerned about ubuntu versions below 18.04?